### PR TITLE
dynamic platform behaviors

### PR DESCRIPTION
**Pull request by danielbe1**
The demo scene is DynamicPlatformDemo.scene and the platform is to the right of the players.

- There are some quirks with the platform when the player rolls under it as a ball it will treat the platform as a solid even when it is semi solid.
- Dropping from the platform isn't perfect yet due to the awkward position change when switching to crouching.

### CODE REVIEW

**Comment by Dejin1** [comment]
Let me know if the platform is working and ready to be reviewed in scene.

File: Assets/Scripts/Common/GameObjectExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+using UnityEngine;
+
+namespace Common
+{
+    public static class GameObjectExtensions
+    {
+        public static T GetOrAdd<T>(this GameObject gameObject) where T : Component
+        {
+            var maybeT = gameObject.GetComponent<T>();
```
Comment (Dejin1) [inline Assets/Scripts/Common/GameObjectExtensions.cs]: maybeT seems like a strange name no need to change was just thinking out loud in a comment lol

File: Assets/Scripts/Platform/Dynamic/BaseDynamicPlatform.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        RefreshConfig();
+    }
+    
+    protected virtual void Start()
+    {
+    }
+    
+    protected virtual void Update()
+    {
+        #if UNITY_EDITOR
```
Comment (Dejin1) [inline Assets/Scripts/Platform/Dynamic/BaseDynamicPlatform.cs]: never seen hash tags on anything other than regions whats the use of the hash tag here just curious?

File: Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+using System;
+using NaughtyAttributes;
```
Comment (Dejin1) [inline Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs]: Did you ever test a build out for an as this attribute causes errors for myself I know you said you may look into was curious if you tested it? If not I will create gifs and send them to you.

File: Assets/Scripts/Common/GameObjectExtensions.cs (hunk 1 above)
Comment (danielbe1) [reply Assets/Scripts/Common/GameObjectExtensions.cs]: it reflects the fact that T might not hold a value (be null)

File: Assets/Scripts/Platform/Dynamic/BaseDynamicPlatform.cs (hunk 2 above)
Comment (danielbe1) [reply Assets/Scripts/Platform/Dynamic/BaseDynamicPlatform.cs]: They are directives for the compiler that are processed before it compiles the actual code. If `UNITY_EDITOR` is not defined in the project (which will happen for non editor builds) this code will be removed.

File: Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs (hunk 3 above)
Comment (danielbe1) [reply Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs]: Yeah, all the builds I made for the old web demo had this attribute in the project. Ill look into it now.

**Comment by danielbe1** [comment]
The platform should be working now.

File: Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs (hunk 3 above)
Comment (danielbe1) [reply Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs]: It wasn't the DLL that caused the problems, it was the weather editor that uses code unavailable on non-editor platforms.

Comment (Dejin1) [reply Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs]: That will be removed soon once I create some prefabs for the weather control your naughty attributes are needed to run your scripts no?

Comment (Dejin1) [reply Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs]: Will your scripts with naughty attributes be able to run if naughty attributes is removed for building the game?

Comment (danielbe1) [reply Assets/Scripts/Platform/Dynamic/DynamicPlatformConfiguration.cs]: Yeah, I will fix the issue with naughty attributes in a separate PR

**Comment by Dejin1** [comment]
looksgood

**Pull request merged**